components/core/src/package/target.rs

```diff
@@ -0,0 +1,197 @@
+}
+
+/// Describes the platform (operating system/kernel)
+/// and architecture (x86_64, i386, etc..) that a package is built for
+#[derive(RustcEncodable, RustcDecodable, Debug, Clone, Hash)]
+pub struct PackageTarget {
+    pub platform: String,
+    pub architecture: String,
+}
+
+impl PackageTarget {
+    /// Creates a new package target
+    ///
+    /// Errors:
+    /// * InvalidPackageTargetPlatform
+    /// * InvalidPackageTargetArchitecture
+    pub fn new<T: Into<String>>(platform: T, architecture: T) -> Self {
+        let valid_platform = PackageTarget::supported_platform(platform).unwrap();
+        let valid_arch = PackageTarget::supported_architecture(architecture).unwrap();
+        PackageTarget {
+            platform: valid_platform,
+            architecture: valid_arch,
+        }
+    }
+
+    fn supported_platform<T: Into<String>>(value: T) -> Result<String> {
+        let platform = value.into();
+        let valid_platforms = vec!["linux", "windows",];
+        if valid_platforms.iter().any(|&x| x == platform) {
+            Ok(platform)
+        } else {
+            Err(Error::InvalidPackageTargetPlatform(platform))
+        }
+    }
+
+    fn supported_architecture<T: Into<String>>(value: T) -> Result<String> {
+        let architecture = value.into();
+        let valid_architectures = vec!["x86_64"];
+        if valid_architectures.iter().any(|&x| x == architecture) {
+            Ok(architecture)
+        } else {
+            Err(Error::InvalidPackageTargetArchitecture(architecture))
+        }
+    }
+}
+
+impl Target for PackageTarget {
+    fn platform(&self) -> &str {
+        &self.platform
+    }
+    fn architecture(&self) -> &str {
+        &self.architecture
+    }
+    fn valid_for(&self, uname: Uname) -> Result<()> {
+        let architecture = uname.machine.trim().to_lowercase();
+        let platform = uname.sys_name.trim().to_lowercase();
+        if self.architecture == architecture && self.platform == platform {
+            Ok(())
+        } else {
+            Err(Error::TargetMatchError(format!("Package target ({}-{}) does not match system \
+                                                 target ({}-{}).",
+                                                self.architecture,
+                                                self.platform,
+                                                architecture,
+                                                platform,)))
+        }
+    }
+}
+
+impl Default for PackageTarget {
+    fn default() -> PackageTarget {
+        let current_system = uname().unwrap();
+        PackageTarget::new(current_system.sys_name.trim().to_lowercase(),
+                           current_system.machine.trim().to_lowercase())
+    }
+}
+
+impl fmt::Display for PackageTarget {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}-{}", self.architecture, self.platform)
+    }
+}
+
+impl AsRef<PackageTarget> for PackageTarget {
```

Comment:
This is the default implementation of `AsRef` which should be automatically available on any type. I don't think you need to provide it here but I could be wrong!